### ISSUE TITLE
Feature request: Disable constructions

### ISSUE THREAD

**Issue by hackedpassword** [yairm210/Unciv#9980]
I tire of fighting with my own cities not to build certain things, and sometimes they sneak it through anyway. Then for say uranium for example, hey where's that resource? It was here a minute ago? Someone prob built a reactor now I have to audit 15 cities and find the perpetrator. Can we denounce our own cities?? Let's do that.

So the proposed feature here is to long press the + (plus) on a construction, any unit, building, wonder, other, whatever. That item is then sent to another section like "Disabled", looking like any other section in construction, still with a plus like any other. If the player hits the plus on the disabled item, pops it back where it should be. Meanwhile, anything in the disabled section is not factored when a city is choosing constructions on its own.

![Screenshot_20230827_105856](https://github.com/yairm210/Unciv/assets/119975434/96884f7a-b63b-49f3-8e43-bf7cc1ff0874)

**Comment by Skekdog** [comment]
You can disable auto city construction in settings

**Comment by hackedpassword** [comment]
I don't want to disable construction, I want to disable certain construction.

**Comment by hackedpassword** [comment]
Is that a hint? Ouch!

**Comment by hackedpassword** [comment]
> 😈 just perpetrating your joke again
LOL you know I got you. 
I'll check that jar out later maybe tonight at best, pressing priorities this day. 

Implementation sounds tricky. I'm guessing then that the lists shown on the construction screen are generated with some inferred data structure magic rather than say a stored list of available construction. Makes sense, consequentially making more sense why a separate blacklist would have to be held. 

Is there a structure for save games, some local per installation config that can associate data to individual saves? Hmm that wouldn't be optimal either when uploading saves. 

I feel this idea may be proven technically impractical seeing the issues involved @SomeTroglodyte.

**Comment by hackedpassword** [comment]
My review of your .jar - a new in-city menu that adds refreshing functionality!

1. I love it. This is a feature that belongs, flows naturally and provides expected gameplay UX.
2. Works as advertised* - I'm tempted to migrate to this .jar for regular games.
3. * As implemented, anything added to the disabled menu is global for all cities. That's a frustration.
4. Disabling all buildings and wonders leaving units enabled will ignore units for auto-build next turn.
5. See #1. Being able to bench a build is way better than I first imagined. It's good.

I hope others give it a try! I'm certain not everyone will feel the same but since you have to long-press/right-click the menu, there's no UI change otherwise.

**Comment by hackedpassword** [comment]
That defeats the purpose - you're trying to prevent your other cities from consuming resources another needs, or, steer a city to a construction goal instead of it auto picking a poor build. Lots more reasons for individual city control.

Really nice work for this demo test! =) more more more LOL

**Comment by hackedpassword** [comment]
[GameSettings.json.txt](https://github.com/yairm210/Unciv/files/12491722/GameSettings.json.txt)

Github doesn't accept json files? lol!

[China - 56 turns.txt](https://github.com/yairm210/Unciv/files/12491737/China.-.56.turns.txt)

Save game. In this state, Bejing produces nothing and maintains producing nothing.

Here's another good one - try removing "Nothing" from the active queue "in all cities". Unciv will be very upset about this.

I added "nothing" to the queue of all cities. Worked... for a little while then Unciv lost coherency and crashed:
D:\Dev\Unciv dev>java -jar Unciv-ConstructionBlacklist.jar
[LWJGL] [ThreadLocalUtil] Unsupported JNI version detected, this may result in a crash. Please inform LWJGL developers.
Exception in thread "main" java.lang.IndexOutOfBoundsException: index can't be >= size: 6 >= 4
        at com.badlogic.gdx.utils.Array.get(Array.java:155)
        at com.unciv.ui.screens.cityscreen.CityConstructionsTable.ensureQueueEntryVisible(CityConstructionsTable.kt:795)
        at com.unciv.ui.screens.cityscreen.CityConstructionsTable.getQueueEntry$selectQueueEntry(CityConstructionsTable.kt:357)
        at com.unciv.ui.screens.cityscreen.CityConstructionsTable.access$getQueueEntry$selectQueueEntry(CityConstructionsTable.kt:71)
        at com.unciv.ui.screens.cityscreen.CityConstructionsTable$getQueueEntry$2.invoke(CityConstructionsTable.kt:362)
        at com.unciv.ui.screens.cityscreen.CityConstructionsTable$getQueueEntry$2.invoke(CityConstructionsTable.kt:362)
        at com.unciv.ui.components.input.ActivationActionMap.activate(ActivationActionMap.kt:56)
        at com.unciv.ui.components.input.ActorAttachments.activate(ActorAttachments.kt:42)
        at com.unciv.ui.components.input.ActivationExtensionsKt.activate(ActivationExtensions.kt:17)
        at com.unciv.ui.components.input.ActivationListener.longPress(ActivationListener.kt:20)
        at com.badlogic.gdx.scenes.scene2d.utils.ActorGestureListener$1.longPress(ActorGestureListener.java:58)
        at com.badlogic.gdx.input.GestureDetector$1.run(GestureDetector.java:58)
        at com.badlogic.gdx.backends.lwjgl3.Lwjgl3Application.loop(Lwjgl3Application.java:208)
        at com.badlogic.gdx.backends.lwjgl3.Lwjgl3Application.<init>(Lwjgl3Application.java:166)
        at com.unciv.app.desktop.DesktopLauncher.main(DesktopLauncher.kt:82)

The general problem with manual queuing comes when your empire is vast and it is no longer practical to micromanage. That's why I'm a big fan of selective automation. As I mentioned in discord, you might have hundreds of troops and a multifront war, I'm going to automate the entirety of say a couple/few unit types and let them run amok while I focus certain units against crucial targets. Meanwhile, I'm not trying to split my attention on if city J and city W are stealing my uranium or building a temple. We're at war people, what are you doing?! Definitely denouncing someone.

fyi, I super really appreciate you taking the time to give this feature a good go @SomeTroglodyte.

Having read how you were assigning to the queue, it makes sense now and I did try your way. So items are assigned then basically from disabled to the queue to isolate construction to that city. Now it makes sense.

That save should demonstrate nothing being superseded with a decision to build.

**Comment by hackedpassword** [comment]
Adding the right click menu in 4.8 is so awesome. Of course in context to this thread, the more challenging aspect didn't make it (yet?) which is not only understandable, it is commendable that the new feature was implemented in a productive way that carries much of the feature with it.

I don't think players have yet to comprehend the addition. I guess that's the norm though I'm this moderately paced environment of updating.

What's the disabled construction feature looking like now? Might that move forward?

**Comment by hackedpassword** [comment]
In my opinion, the matter that began this thread is the benefit goal. It remains a constant struggle with my own empire to handle themselves conducively to war objectives, not oooo I can build X yea that's a great option! without having consulted the strategic interests of other cities not the emporer/ss themselves. Wait, are we emporers? Has that ever been addressed?

btw the logical process described above would be a nice AI addition, verbalizing the thought (not a request). 

I suppose another way to approach this could be flagging a do-not-build constructor, negative connotation where an internal list is referenced and any city that hasn't been told to build X, don't.

Per city has it's benefits also. When not caring about how other cities are going about their business, then you have some new cities and/or recently annexed cities and you want those to focus away from say culture or spirituality buildings, drop those from being queued and check back on them in 20-30 turns cause they're not really contributing anything to the civ at this point.

I started this thread for the former, so that's my vote. I see the second option as viable also, so both features would be great additions. If limited to one or the other, I'd point that call at the community.

**Comment by hackedpassword** [comment]
> having consulted the strategic interests of other cities not the emporer/ss themselves
This is an open idea towards better AI.

> hard to parse
A bit scattered, yes, verbalizing rationale can do that lol

> works as intended
Hmm a formula, ok! Since the city/s have decided to produce nothing, would it be reasonable to produce a line under Cities in Notifications, something like "[city] reports no units need producing."? This seems similar to when terrain destruction occurred silently due to units configured to destroy terrain when attacking - a notification was included to convey what occurred removing confusion as to why the mystery destruction. Useful civ telemetry.

**Comment by hackedpassword** [comment]
> multiplayer

WHAT? LMAO how would that even work?? I've never played unciv in multiplayer so I'll go with whatever. That concept amuses me to no end. Players better pay close attention.... 

We'll see how this works in-game when published. I'll thank/blame/applaud/complain trying the results of said madness. Ok not madness but useful exclusion policy. Looking forward to it. :D

**Comment by hackedpassword** [comment]
@SomeTroglodyte

Your implementation of blacklisting is beautiful in execution.

Step 1: Disable construction of building/unit/wonder - > all cities respond, no one builds
Step 2: Individual cities then may select to construct
Step 3: Reestablish build across empire when desired

It looks good, feels good, plays good. :100:

**Comment by hackedpassword** [comment]
This may or may not be intended - using the add to all cities option includes puppet cities. That supersedes their queue, when usually their queue is out of the player's control.

**Comment by hackedpassword** [comment]
I love this menu.

**Comment by hackedpassword** [comment]
@SomeTroglodyte fyi found a problematic situation.

258 turns in, discovered a half dozen buildings under Disabled left over from the previous game that my civ should have been working on.

The immediate problem here is that for the rest of this game, I'll have to monitor the Disabled menu for unexpected disabled buildings.

However, a workaround will be to start a new game in the Future era, clear the menu, return to the current game.
